File: im-mobile/lib/services/screenshot_notification_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/screenshot_notification.dart';

class ScreenshotNotificationService {
  static const String _baseUrl = '/api/screenshot';

  Future<ScreenshotEvent?> reportScreenshot(int conversationId, String conversationType) async {
    final userId = await _getUserId();
    final username = await _getUsername();

    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/report'),
        headers: {
          'Content-Type': 'application/json',
          'X-User-Id': '$userId',
          'X-Username': username,
        },
        body: jsonEncode({
          'conversationId': conversationId,
          'conversationType': conversationType,
          'deviceType': 'mobile',
          'deviceInfo': 'Flutter Mobile',
        }),
      );

      if (response.statusCode == 200) {
        return ScreenshotEvent.fromJson(jsonDecode(response.body));
      }
    } catch (e) {
      // Silently fail
    }
    return null;
  }

  Future<ScreenshotSettings> getSettings() async {
    final userId = await _getUserId();
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/settings'),
        headers: { 'X-User-Id': '$userId' },
      );
      if (response.statusCode == 200) {
        return ScreenshotSettings.fromJson(jsonDecode(response.body));
      }
    } catch (e) {
      // Return default
    }
    return ScreenshotSettings(userId: userId);
  }

  Future<ScreenshotSettings> updateSettings(ScreenshotSettings settings) async {
    final response = await http.put(
      Uri.parse('$_baseUrl/settings'),
      headers: {
        'Content-Type': 'application/json',
        'X-User-Id': '${settings.userId}',
      },
      body: jsonEncode(settings.toJson()),
    );
    return ScreenshotSettings.fromJson(jsonDecode(response.body));
  }

  Future<List<ScreenshotEvent>> getHistory({int limit = 50}) async {
    final userId = await _getUserId();
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/history?limit=$limit'),
        headers: { 'X-User-Id': '$userId' },
      );
      if (response.statusCode == 200) {
        final List<dynamic> list = jsonDecode(response.body);
        return list.map((e) => ScreenshotEvent.fromJson(e)).toList();
      }
    } catch (e) {
      // Return empty
    }
    return [];
  }

  Future<int> _getUserId() async {
    return 1;
  }

  Future<String> _getUsername() async {
    return 'User';
  }
}

File: im-desktop/src/services/screenshot-notification-service.ts
import { ScreenshotEvent, ScreenshotSettings, ScreenshotEventRequest } from '../types/screenshot-notification';

const API_BASE = '/api/screenshot';

class ScreenshotNotificationService {
  private static instance: ScreenshotNotificationService;

  static getInstance(): ScreenshotNotificationService {
    if (!ScreenshotNotificationService.instance) {
      ScreenshotNotificationService.instance = new ScreenshotNotificationService();
    }
    return ScreenshotNotificationService.instance;
  }

  async reportScreenshot(conversationId: number, conversationType: string): Promise<ScreenshotEvent | null> {
    const userId = Number(localStorage.getItem('userId'));
    const username = localStorage.getItem('username') || 'Unknown';

    const request: ScreenshotEventRequest = {
      conversationId,
      conversationType: conversationType as 'private' | 'group',
      deviceType: 'desktop',
      deviceInfo: navigator.userAgent,
    };

    try {
      const response = await fetch(`${API_BASE}/report`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-User-Id': String(userId),
          'X-Username': username,
        },
        body: JSON.stringify(request),
      });
      if (!response.ok) return null;
      return await response.json();
    } catch (error) {
      console.error('Failed to report screenshot:', error);
      return null;
    }
  }

  async getSettings(): Promise<ScreenshotSettings> {
    const userId = Number(localStorage.getItem('userId'));
    try {
      const response = await fetch(`${API_BASE}/settings`, {
        headers: { 'X-User-Id': String(userId) },
      });
      return await response.json();
    } catch {
      return { userId, enableScreenshotNotification: true, notifyOnCapture: true, receiveScreenshotAlerts: true, alertForContacts: true, alertForGroups: true, silentMode: false };
    }
  }

  async updateSettings(settings: Partial<ScreenshotSettings>): Promise<ScreenshotSettings> {
    const userId = Number(localStorage.getItem('userId'));
    const response = await fetch(`${API_BASE}/settings`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json', 'X-User-Id': String(userId) },
      body: JSON.stringify({ ...settings, userId }),
    });
    return await response.json();
  }

  async getHistory(limit = 50): Promise<ScreenshotEvent[]> {
    const userId = Number(localStorage.getItem('userId'));
    const response = await fetch(`${API_BASE}/history?limit=${limit}`, {
      headers: { 'X-User-Id': String(userId) },
    });
    return await response.json();
  }
}

export default ScreenshotNotificationService;
